商品出品機能　実装

tamtam-time opened the pull request:
以下、確認用画像になります。
https://gyazo.com/da7237f7f8e961c98ce0c5ff36b9cf10

https://gyazo.com/469d02e036b1ee66026975f757831da5

https://gyazo.com/b4cc781b4df9c5e93d5f9074b4a3687e

https://gyazo.com/360970b74f2428018ee3298a45f0b6c9

https://gyazo.com/d99e74bd0235412dcfd5146a0f47bc0b

https://gyazo.com/3069bb6ada58f8b0aaa9ce6e128f5701

宜しくお願いいたします。

tamtam-time committed:
商品出品機能　実装

以下、確認用画像になります。
https://gyazo.com/da7237f7f8e961c98ce0c5ff36b9cf10

https://gyazo.com/469d02e036b1ee66026975f757831da5

https://gyazo.com/b4cc781b4df9c5e93d5f9074b4a3687e

https://gyazo.com/360970b74f2428018ee3298a45f0b6c9

https://gyazo.com/d99e74bd0235412dcfd5146a0f47bc0b

https://gyazo.com/3069bb6ada58f8b0aaa9ce6e128f5701

宜しくお願いいたします。

tamtam-time commented:
# What
商品の出品機能

# Why
商品出品機能の実装

div-reviewer reviewed (commented):
tamtam-time様、学習お疲れさまです。
お忙しい中商品出品機能のプルリクエスト提出ありがとうございました。

挙動動画・画像のご提出もありがとうございます。
今回、リンクの上部に説明がなかったのでお手数ですが追加後、再度googleフォームでプルリクエストの提出をお願いできますでしょうか。

以下の実装条件を満たすため：
```ruby
・Gyazoのリンクを添付する際は、何を撮影したGyazoのリンクか分かるように、リンクの前に説明を記載すること。
```

urlを貼り付けて頂いた枠の右上にある三点リーダー（…）から変更可能になります。
操作についてご不明点ございましたらチャット・コネクトメンターにお気軽にご相談下さい。

どうぞ宜しくお願い致します。

tamtam-time commented:
> 以下、確認用画像になります。

ログイン状態の場合は、商品出品ページへ遷移できる動画
 https://gyazo.com/da7237f7f8e961c98ce0c5ff36b9cf10

>価格が入力されると同時に、販売手数料と販売利益が表示される動画
> https://gyazo.com/469d02e036b1ee66026975f757831da5

> 必要な情報を適切に入力して「出品する」ボタンを押すと、商品情報がデータベースに保存される動画
> https://gyazo.com/b4cc781b4df9c5e93d5f9074b4a3687e

> 入力に問題がある状態で「出品する」ボタンが押された場合、情報は保存されず、出品ページに戻りエラーメッセージが表示される動画
> https://gyazo.com/360970b74f2428018ee3298a45f0b6c9

> ログアウト状態の場合は、商品出品ページへ遷移しようとすると、ログインページへ遷移する動画
> https://gyazo.com/d99e74bd0235412dcfd5146a0f47bc0b

> テスト結果の画像
> https://gyazo.com/3069bb6ada58f8b0aaa9ce6e128f5701
> 
> 宜しくお願いいたします。

div-reviewer commented on app/controllers/items_controller.rb:
```diff
@@ -1,20 +1,33 @@
+  before_action :authenticate_user!, except: [:index, :show]
 
   def index
+    @item = Item.includes(:user).order('created_at DESC')
```
```ruby
 @item = Item.includes(:user).order('created_at DESC')
```

上記の文は削除かコメントアウトしましょう。

**理由**
こちらの記述は次のフェーズである「商品一覧機能」の実装範囲となります。
今コードを記述してしまいますと、商品一覧機能の実装時に差分として確認できなくなる為、商品出品フェーズの間ではコメントアウトもしくは削除しておきましょう。

div-reviewer commented on app/controllers/items_controller.rb:
```diff
@@ -1,20 +1,33 @@
+  def show
 
-  #end
+  end
```
```ruby
def show
end
```

こちらは詳細の実装で作成するアクションになります。
現段階では特に使用しない為、コメントアウトのままにしましょう。

以下の部分も合わせて修正しましょう。
```ruby
 before_action :authenticate_user!, except: [:index, :show]
↓
例）
 before_action :authenticate_user!, except: [:index] #, :show]
```

div-reviewer commented on app/javascript/packs/application.js:
```diff
@@ -3,6 +3,19 @@
+     addTaxDom.innerHTML = Math.round(priceInput.value * 0.1 );
+     const addPofitDom = document.getElementById("profit");
+     // console.log(priceInput.innerHTML);
+     addPofitDom.innerHTML = Math.round(priceInput.value - Math.round(priceInput.value * 0.1 ))
```
```ruby
 //  console.log(priceInput)
// console.log(priceInput.innerHTML);
// console.log(priceInput.innerHTML);
```

デバッグで実装が仕様通りできているか確認ができていましたね。素晴らしいです。
console.logやbinding.pryといったデバック用コードは削除しておきましょう。

**理由**
デバッグ用のコードは開発中に使用するものであるためです。
コードレビューは一旦完成したコードに対して行うものとお考えください。
もし学習のメモとして残したい場合は、第3者にも分かるように「学習メモ」など補足を付け足しましょう。

div-reviewer commented on app/javascript/packs/application.js:
```diff
@@ -3,6 +3,19 @@
 // a relevant structure within app/javascript and only use these pack files to reference
 // that code so it'll be compiled.
 
```
**①価格用のjsファイルを作成し、内容を転記しましょう。**

購入機能でクレジットカード決済をする際にjsを作成する為、合わせて記載すると、1ファイルのコードの量が多くなってしまいます。
ファイルを分けて可読性をあげましょう。

■参考
コース：最終課題
タイトル：実装に悩んだときのヒント〜商品出品機能〜
STEP2 金額入力時に販売手数料や販売利益の計算結果が表示されるようにする手順を確認しよう

**②Math.round（四捨五入）ではなく、Math.floor（切り捨て）を使用しましょう。**

以下の実装条件を満たすためです。
```ruby
・販売手数料と販売利益は、`小数点以下を切り捨てて`表示すること。
```

div-reviewer commented on app/models/category.rb:
```diff
@@ -0,0 +1,18 @@
+class Category < ActiveHash::Base
+  self.data = [
+    { id: 0, name: '--' },
```
```ruby
{ id: 0, name: '--' },
```

細かい部分ですが、選択していないデータは-が3つになりますので修正しましょう。

■補足
以下のファイルも一緒に修正をお願いいたします。
・app/models/postage.rb
・app/models/sendday.rb
・app/models/status.rb

div-reviewer commented on app/models/item.rb:
```diff
@@ -0,0 +1,40 @@
+    validates :send_day_id
+    # 300円以上かつ9,999,999円以下で、半角数字でないと入力不可
+    validates :price, numericality: { greater_than_or_equal_to: 300, less_than_or_equal_to: 9_999_999 }
+  end
```
```ruby
    validates :price, numericality: { greater_than_or_equal_to: 300, less_than_or_equal_to: 9_999_999 }
  end
```

金額のバリデーション設置ありがとうございました。中身をもう一度確認してみましょう。

**■理由**
```ruby
only_integer: trueがないと小数の販売価格でもバリデーションが通ってしまう為

例）
numericality: { only_integer: true, greater_than_or_equal_to: 300, less_than_or_equal_to: 9_999_999 }
```

**■参考**
```ruby
コース：最終課題
タイトル：最終課題補足カリキュラム〜1つのフォームから複数のテーブルに情報を保存するアプリを作成しよう〜

バリデーションを設定しましょう
```

div-reviewer commented on app/models/item.rb:
```diff
@@ -0,0 +1,40 @@
+  has_one_attached :image
+
+  with_options presence: true do
+    validates :user_id
```
```ruby
  with_options presence: true do
    validates :user_id
```

上記の「validates :user_id」、こちらの記述は不要となりますので削除しておきましょう。

**理由**
```ruby
belongs_to :user
```
こちらのアソシエーションの記述によりuserのデータは既に必須になっている為です。

■参考
```ruby
コース：応用
タイトル：RSpec〜PicTweetのモデルの単体テストコードを書こう〜

exampleを整理しよう
```

div-reviewer commented on app/views/items/new.html.erb:
```diff
@@ -5,10 +5,10 @@
     <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
-    <%# render 'shared/error_messages', model: f.object %>
+    <%= render 'shared/error_messages', model: f.object %>
     <%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
```ruby
<%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
<%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```

実装`済みの`案内用文は削除しましょう。

**理由**
コードレビューは一旦完成したコードに対して行うものであるのと、企業様に提出する際に実装案内のコメントが残っていることは好ましくないためです。
もし学習メモとして必要であれば、第三者でもわかるように「残す目的」「なぜ必要か」分かるよう、`「学習メモ」`等分かりやすいコメントとして書いておくことをおすすめいたします。

■補足
他にも不要な案内文があれば、削除しましょう。

div-reviewer commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,127 @@
+
+RSpec.describe Item, type: :model do
+  before do
+    @item = FactoryBot.create(:item)
```
```ruby
@item = FactoryBot.create(:item)
```

こちら[ユーザー管理機能のテスト同様](https://github.com/tamtam-time/furima-38599/pull/11/files#diff-bb8bcb6ef2e13ab9491f4f9ec48be47ee025d7f0752a25029d40810af2c65ee1)、createではなくbuildで作成しましょう。

テスト用のデータベースにテストを行うたびに毎回データが作成され負荷がかかるのを防ぐためです。

div-reviewer commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,127 @@
+      it 'ユーザー登録している人でないと出品できない' do
+        @item.user_id = nil
+        @item.valid?
+        expect(@item.errors.full_messages).to include('User must exist', "User can't be blank")
```
```ruby
 @item.user_id = nil
```

今回モデルに記載いただいているuser_idに対する必須条件を削除して頂くと、
こちらの記載も少々変更になるかと思われます。

以下のカリキュラムを参考に修正してみましょう。

■参考
コース：応用
タイトル：RSpec〜PicTweetのモデルの単体テストコードを書こう〜
 問題13：「ユーザーが紐付いていなければ投稿できない」についてのテストコードを書きましょう

div-reviewer commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,127 @@
+        @item.price = 10_000_000
+        @item.valid?
+        expect(@item.errors.full_messages).to include('Price must be less than or equal to 9999999')
+      end
```
販売価格については、モデルのバリデーションで「半角数字」である制限を記載している為、以下4点のテストを行い、確認していきましょう。
※済とついているものは既に実施済みですので追加の必要はございません。

```ruby
・価格が空では出品できない【済】
・価格に半角数字以外が含まれている場合は出品できない（※半角数字以外が一文字でも含まれていれば良い）
・価格が300円未満では出品できない【済】
・価格が9_999_999円を超えると出品できない【済】
```

div-reviewer commented on app/controllers/application_controller.rb:
```diff
@@ -1,6 +1,6 @@
   before_action :basic_auth
-  # before_action :configure_permitted_parameters, if: :devise_controller?
+  before_action :configure_permitted_parameters, if: :devise_controller?
 
```
```ruby
    devise_parameter_sanitizer.permit(:sign_up, keys: [:nickname, :email, :password, :encrypted_password, :first_name, :last_name, :first_name_yomi, :last_name_yomi, :birthday,])
```

前回の範囲の指摘となってしまい申し訳ございません。
devise_parameter_sanitizer.permit(:sign_up, keys: [〜])の中にはデフォルトで備わっているメールとパスワードは記載なくて問題ございませんので削除しましょう。

■参考
コース：応用
タイトル：Ruby on Rails（ChatApp）〜ChatAppのユーザー管理機能を実装しよう〜
ユーザーの名前をDBへ保存できるように設定しましょう

div-reviewer reviewed (commented):
tamtam-time様、学習お疲れ様です！
お忙しい中商品出品の再レビュー依頼ありがとうございます。挙動動画・画像とプルリクエスト確認させていただきました。

with_optionsを使用してバリデーションがまとめられているなど、可読性を考えたコーディングが意識できていて素晴らしいです。この調子で最後まで頑張りましょう。
修正点をコメントさせて頂きましたので、ご確認をお願い致します。

```ruby
【次回プルリクエスト提出する際の注意点】
①修正が終わりましたら、`コミットとプッシュ`をして、`プルリクエストに修正内容が反映されているか確認`しましょう。
②既に頂いている挙動確認の画像や動画が修正範囲だった場合、最新の画像や動画の提出をお願い致します。
③googleフォームの送信をお願いいたします。
```

ご不明点や疑問点ございましたら、チャット・コネクトメンターにご相談ください！
どうぞよろしくおねがいします。

tamtam-time committed:
商品出品機能　修正分

テスト修正しましたので、テスト結果の画像を以下に記載いたします。
https://gyazo.com/59e040bc19175e12e031282dcf15565a

ShotaOtake commented on app/models/item.rb:
```diff
@@ -0,0 +1,33 @@
+    validates :postage_id
+    validates :area_id
+    validates :send_day_id
+    validates :price, numericality: { greater_than_or_equal_to: 300, less_than_or_equal_to: 9_999_999 }
```
◎バリデーションを修正しましょう！

```
validates :price, numericality: { greater_than_or_equal_to: 300, less_than_or_equal_to: 9_999_999 }
```

→現在の記述では小数の販売価格での登録が可能な状態となっているため。

ShotaOtake reviewed (commented):
レビュー依頼を頂きありがとうございます！
追加で修正依頼のコメントをさせて頂きましたので、ご確認の上修正をお願い致します。

完成まであと一歩です！
引き続き実装頑張っていきましょう！

```
※テストコードにて修正箇所のございます際は、お手数ですが、最新の実行結果の画像も改めて添付頂きますようお願い致します。
```

tamtam-time committed:
商品出品機能　修正

前回テストコード修正し、画像URL掲載しておりましたが、今回はテストコード修正しておりませんが、念の為、最新のテスト結果のURLを記載いたします。
https://gyazo.com/c8ef3cfc856855548a54799bc22eac4c

宜しくお願いいたします。

AsukaToyoshima reviewed (commented):
レビューのご依頼ありがとうございます。
こちらLGTMとさせていただきます。お疲れ様でした！

次のブランチを切る前に、本プルリクエストのMergeとMasterブランチへのPullを忘れないようにしておきましょう。

引き続き、実装頑張っていきましょう！　応援しております！
